### PR TITLE
Refinance test

### PR DESCRIPTION
**Pull request by paszkowskiDamian**

### CODE REVIEW

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts [hunk 1]
```diff
@@ -23,23 +25,49 @@ export class Simulator<
   }
 
   static create<S extends SimulationStrategy>(schema: S) {
-    return new Simulator(schema)
+    return new Simulator(schema, schema)
```
Comment (piotrwitek) [inline sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:28]: is this a mistake?

File: sdk/simulator-service/tests/simulator.test.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        toToken: Token
+    }) => {
+        return {
+            provider: SwapProviderType.OneInch,
+            fromTokenAmount: params.fromAmount,
+            toTokenAmount: TokenAmount.createFrom({ token: params.toToken, amount: '10' }),
+            estimatedGas: '0'
+        }
+    }
+}
```
Comment (piotrwitek) [inline sdk/simulator-service/tests/simulator.test.ts:92]: such mock could be extracted and shared betwoon tests

File: sdk/simulator-service/tests/simulator.test.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+            expect(targetPosition.debtAmount).toEqual(testSourcePosition.debtAmount)
+        })
+
+        it('should open position as required target pool', async () => {
+            const targetPosition = simulation.targetPosition
+
+            expect(targetPosition.pool).toEqual(testTargetLendingPool)
+        })
+    })
```
Comment (piotrwitek) [inline sdk/simulator-service/tests/simulator.test.ts:137]: please also assert if positionId is not null and correct shape

**Review by piotrwitek — CHANGES_REQUESTED** [review]

**Comment by piotrwitek** [comment]
Looks great, my only request is to add assertions for target positionId. Hope that's useful :)

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts (hunk 1 above)
Comment (paszkowskiDamian) [reply sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:28]: No it's not, I realized that I need to keep the original schema for processing. It works in such way that when adding new step it removes first step from schema for typing purposes I wanted Simulator to be a generic that with Schema array being reduced each time you add next step.

I guess I could have done it differently, and just keep the schema and index of current processing step.

File: sdk/simulator-service/tests/simulator.test.ts (hunk 2 above)
Comment (paszkowskiDamian) [reply sdk/simulator-service/tests/simulator.test.ts:92]: ok 👍

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts (hunk 1 above)
Comment (robercano) [reply sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:28]: Do you mind adding a comment to that? It's going to raise some questions in the future I'm afraid. We can talk about it when you are back. Thanks!

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts [hunk 4]
```diff
@@ -23,23 +25,49 @@ export class Simulator<
...
   }
 
   public async run(): Promise<SimulationState> {
-    const getReference = (path: [string, string]) => {
-      const [step, name] = path
-      return {
+    for (let i = 0; i < this.nextArray.length; i++) {
+      const proccesesedStepSchema = this.originalSchema[i]
```
Comment (robercano) [inline sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:33]: Should we check that `nextArray` and `originalSchema` have the same number of elements?

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts [hunk 5]
```diff
@@ -23,23 +25,49 @@ export class Simulator<
...
+      const getReference = (path: [string, string]) => {
+        const [stepName, output] = path
+        const step: steps.Steps | undefined = this.state.steps[stepName]
```
Comment (robercano) [inline sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:36]:
```suggestion
        const step: Maybe<steps.Steps> = this.state.steps[stepName]
```

File: sdk/simulator-service/tests/simulator.test.ts
Comment (robercano) [inline sdk/simulator-service/tests/simulator.test.ts:1]: 👏

**Review by robercano — APPROVED** [review]
Looks good to me! I've added some comments and suggestions. I'll approve anyway, but please let me know if you are not going to address some of them

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts (hunk 4 above)
Comment (paszkowskiDamian) [reply sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:33]: no as nextArray gets first element removed with next step added.

Comment (paszkowskiDamian) [reply sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:33]: so original schema will always be at least the lenght of next Array.

File: sdk/simulator-service/src/implementation/simulator-engine/simulator.ts (hunk 1 above)
Comment (paszkowskiDamian) [reply sdk/simulator-service/src/implementation/simulator-engine/simulator.ts:29]: I can refactor it one I'm back

**Review by piotrwitek — DISMISSED** [review]

**Review by BackToTheCode — APPROVED** [review]
Reapproving

**Pull request merged**